internal/provider/data_source.go

```diff
@@ -31,105 +48,111 @@ func dataSource() *schema.Resource {
 			"or external programs beyond standard shell utilities, so it is not recommended to use this data source " +
 			"within configurations that are applied within Terraform Enterprise.",
 
-		ReadWithoutTimeout: dataSourceRead,
-
-		Schema: map[string]*schema.Schema{
-			"program": {
+		Attributes: map[string]schema.Attribute{
+			"program": schema.ListAttribute{
 				Description: "A list of strings, whose first element is the program to run and whose " +
 					"subsequent elements are optional command line arguments to the program. Terraform does " +
 					"not execute the program through a shell, so it is not necessary to escape shell " +
 					"metacharacters nor add quotes around arguments containing spaces.",
-				Type:     schema.TypeList,
-				Required: true,
-				Elem: &schema.Schema{
-					Type: schema.TypeString,
+				ElementType: types.StringType,
+				Required:    true,
+				Validators: []validator.List{
+					listvalidator.SizeAtLeast(1),
 				},
-				MinItems: 1,
 			},
 
-			"working_dir": {
+			"working_dir": schema.StringAttribute{
 				Description: "Working directory of the program. If not supplied, the program will run " +
 					"in the current directory.",
-				Type:     schema.TypeString,
 				Optional: true,
-				Default:  "",
 			},
 
-			"query": {
+			"query": schema.MapAttribute{
 				Description: "A map of string values to pass to the external program as the query " +
 					"arguments. If not supplied, the program will receive an empty object as its input.",
-				Type:     schema.TypeMap,
-				Optional: true,
-				Elem: &schema.Schema{
-					Type: schema.TypeString,
-				},
+				ElementType: types.StringType,
+				Optional:    true,
 			},
 
-			"result": {
+			"result": schema.MapAttribute{
 				Description: "A map of string values returned from the external program.",
-				Type:        schema.TypeMap,
+				ElementType: types.StringType,
+				Computed:    true,
+			},
+
+			"id": schema.StringAttribute{
+				Description: "The id of the data source. This will always be set to `-`",
 				Computed:    true,
-				Elem: &schema.Schema{
-					Type: schema.TypeString,
-				},
 			},
 		},
 	}
 }
 
-func dataSourceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+func (n *externalDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var config externalDataSourceModelV0
 
-	programI := d.Get("program").([]interface{})
-	workingDir := d.Get("working_dir").(string)
-	query := d.Get("query").(map[string]interface{})
+	diags := req.Config.Get(ctx, &config)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	var program []string
```

Comment:
If there's not already an acceptance test verifying `null` values in the `program` list, it might be good to add one. e.g.

```terraform
data "external" "test" {
  program = ["...", null, "..."]
}
```

I think we may want to err on the side of caution here and use `[]types.String` here, then filter out elements that match `prgramArgRaw.IsNull() || programArgRaw.ValueString() == ""`

We could potentially consider adding explicit validation in the schema to prevent null elements in a future major version.